lib/Conversion/TorchToLinalg/Uncategorized.cpp

```diff
@@ -1015,6 +1015,56 @@ static Value createLinalgPayloadCalculationForElementwiseOp(
     }
     return result;
   }
+  if (auto clampTensor = dyn_cast<AtenClampTensorOp>(op)) {
+    AtenClampTensorOp::Adaptor adaptor(operands);
+    auto min = adaptor.getMin();
+    auto max = adaptor.getMax();
+    if (min.getType().isa<Torch::OptionalType>() ||
+        max.getType().isa<Torch::OptionalType>()) {
+      clampTensor.emitError("unimplemented: runtime optional type");
+      return nullptr;
+    }
+    Type dtype = converter->convertType(clampTensor.getType())
+                     .cast<RankedTensorType>()
+                     .getElementType();
+    bool isMinNone = true;
+    auto result = payloadArgs[0];
+    if (!min.getType().isa<Torch::NoneType>()) {
+      isMinNone = false;
+      auto minPromoted = convertScalarToDtype(b, loc, payloadArgs[1], dtype);
+      Value pred;
+      if (dtype.isa<mlir::FloatType>()) {
+        pred = b.create<arith::CmpFOp>(loc, arith::CmpFPredicate::ULT, result,
+                                       minPromoted);
+      } else if (dtype.isa<mlir::IntegerType>()) {
+        pred = b.create<arith::CmpIOp>(loc, arith::CmpIPredicate::slt, result,
+                                       minPromoted);
+      } else {
+        clampTensor.emitError(
```

Comment:
Missing return?

I think this function needs to be refactored. It is very important that:

A. Any failure returns happen before any ops are created. Pattern match predictes should be at the top.
B. The pattern be made resilient to multiple application and partial application.

We really shouldn't be emitting errors during pattern application like this due to the above. We can notifyMatchResult or mutate to at an attribute indicating what is wrong (with a pass at the end to validate and report errors.

Since this kind of thing is already being done here, I'm not going to request that you did all error emission in this patch, but could you please raise an issue to refactor out all error reporting from these patterns?

You do need to move all checks that will return failure to the top before creating any ops though. I expect that if you have a unit test that tests this failure case, it will crash as written.

Reply:
Opened an issue here:https://github.com/llvm/torch-mlir/issues/2627